.github/scripts/download-node-tooling.sh

```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+
+abort() {
+  echo $1
+  exit 1
+}
+
+download_file() {
+    local url=$1
+    local target=$2
+
+    echo "Downloading $(basename "$url")..."
+    mkdir -p "$(dirname "$target")"
+    curl --silent --fail --show-error --retry 3 --retry-delay 30 --output "$target" "$url"
+}
+
+detect_platform() {
+    local platform="$(uname -s | tr '[:upper:]' '[:lower:]')"
+
+    case "${platform}" in
+      linux) platform="linux" ;;
+      darwin) platform="darwin" ;;
+      windows) platform="win" ;;
+    esac
+
+    echo "${platform}"
+}
+
+detect_arch() {
+    local arch="$(uname -m | tr '[:upper:]' '[:lower:]')"
+
+    case "${arch}" in
+      x86_64 | amd64) arch="x64" ;;
+      arm64 | aarch64) arch="arm64" ;;
+    esac
+
+    echo "${arch}"
+}
+
+node_url() {
+    local version=$1
+    local platform="$(detect_platform)"
+    local arch="$(detect_arch)"
+
+    echo "https://nodejs.org/dist/v$version/node-v$version-$platform-$arch.tar.gz"
+}
+
+pnpm_url() {
+    local version=$1
+
+    echo "https://registry.npmjs.org/pnpm/-/pnpm-$version.tgz"
+}
+
+main() {
+    local node_version=$1
+    local pnpm_version=$2
+
+    if [ "$node_version" == "" ] || [ "$pnpm_version" == "" ]; then
+        abort "Usage: download-node-tooling.sh <NODE_VERSION> <PNPM_VERSION>"
+    fi
+
+    local target_directory=~/.m2/repository/com/github/eirslett
+    local platform="$(detect_platform)"
+    local arch="$(detect_arch)"
+    local ext="$([ "$arch" == "win" ] && echo "zip" || echo "tar.gz")"
+
+    download_file "$(node_url "$node_version")" "$target_directory/node/$node_version/node-$node_version-$platform-$arch.$ext"
+    download_file "$(pnpm_url "$pnpm_version")" "$target_directory/pnpm/$pnpm_version/pnpm-$pnpm_version.$ext"
```

Comment:
This would mean that we'd have to invoke Maven, which would download a bunch more artifacts. The logic inside of the plugin is error prone, which is why we want to circumvent it here entirely to avoid issues.

Reply:
The plugin's download paths and filenames are part of an internal API (and they're not actually supposed to be downloading arbitrary stuff into the local maven repo like that, so there could be a change in the future), so if the plugin changes the paths on you, it will silently break this caching script. The plugin won't find your cached stuff because it's looking at a new path, and resume the old behaviour of trying to download at build time. So, I still believe that replicating the download with curl is more error prone than just using the plugin itself in a retry loop.

If you really don't want to use maven call here, perhaps you could do the same thing the product build does: Just tell frontend to use system NPM/PNPM? The product build can't download stuff from the internet, so this is already a solved problem there

To that you would add `-Dskip.installnodepnpm -Dskip.installnodenpm` to the maven arguments, and then before the build populate the build directory with what frontend plugin needs:

```bash
pnpm_version="$(sed -r 's,^.*<pnpm.version>([^<]+)</pnpm.version>.*$,\1,;tx;d;:x' pom.xml)"
npm install -g "pnpm@$pnpm_version"

mkdir -p "node"
ln -s /usr/lib/node_modules/ "node/node_modules"
ln -s /usr/bin/node "node/node"
```

Reply:
I am aware that the paths should not be considered as stable, but that doesn't matter as the impact of this missing the cache would simply be an increase in workflow failures, which we can then fix when it happens. Nothing critical or user facing.

Another reason why we cannot use the Maven plugin directly is because it will download OS specific binaries, and the build is executed only under Linux (`ubuntu-latest`). So it would not be able to fill the cache with all the relevant binaries.

> Just tell frontend to use system NPM/PNPM? The product build can't download stuff from the internet, so this is already a solved problem there

I am honestly starting to lean towards this, that would allow us to just use the GitHub action for setting up Node.js, which saves a lot of headache. Let me do some experimenting.

Reply:
After doing some digging I think we just want to stick with the script solution for now, we can always improve this later.

Reply:
Created a seperate PR to handle this futher: https://github.com/keycloak/keycloak/pull/34984